linux/ibus-keyman/src/engine.c

```diff
@@ -576,56 +631,150 @@ ibus_keyman_engine_process_key_event (IBusEngine     *engine,
+                            g_free(keyman->char_buffer);
+                            g_free(utf8_pair);
+                            keyman->char_buffer = new_buffer;
+                        }
+                        keyman->firstsurrogate = 0;
+                    }
                 }
-                if (numbytes)
-                {
-                    utf8[numbytes] = 0;
+                else {
+                    gchar *utf8 = (gchar *) g_new0(gchar, 12);
+                    numbytes = g_unichar_to_utf8(action_items[i].character, utf8);
+                    if (numbytes > 12) {
+                        g_error("g_unichar_to_utf8 overflowing buffer");
+                        g_free(utf8);
+                    }
+                    else {
+                        g_message("unichar:U+%04x, bytes:%d, string:%s", action_items[i].character, numbytes, utf8);
+                        if (keyman->char_buffer == NULL) {
+                            keyman->char_buffer = utf8;
+                        }
+                        else {
+                            gchar *new_buffer = g_strjoin("", keyman->char_buffer, utf8, NULL);
+                            g_free(keyman->char_buffer);
+                            g_free(utf8);
+                            keyman->char_buffer = new_buffer;
+                        }
+                    }
                 }
-                g_message("CHAR action unichar:U+%04x, bytes:%d, string:%s", action_items[i].character, numbytes, utf8);
-                ibus_keyman_engine_commit_string(keyman, utf8);
                 break;
             case KM_KBP_IT_MARKER:
+                keyman->hasonlymarker = TRUE;
                 g_message("MARKER action");
                 break;
             case KM_KBP_IT_ALERT:
                 g_message("ALERT action");
+                Display * alert_display  = XOpenDisplay(NULL);
+                XkbBell(alert_display, None, 0, None);
+                XCloseDisplay(alert_display);
                 break;
             case KM_KBP_IT_BACK:
                 g_message("BACK action");
-                g_message("DAR: ibus_keyman_engine_process_key_event - client_capabilities=%x, %x", engine->client_capabilities,  IBUS_CAP_SURROUNDING_TEXT);
-
-                if ((engine->client_capabilities & IBUS_CAP_SURROUNDING_TEXT) != 0) {
-                    g_message("deleting surrounding text 1 char");
-                    ibus_engine_delete_surrounding_text(engine, -1, 1);
-                } else {
-                    g_message("forwarding backspace");
-                    forward_keycode(keyman, Keyman_Pass_Backspace_To_IBus, 0);
+                if (keyman->char_buffer != NULL)
+                {
+                    ibus_keyman_engine_commit_string(keyman, keyman->char_buffer);
+                    g_free(keyman->char_buffer);
+                    keyman->char_buffer = NULL;
+                }
+                if (keyman->hasonlymarker) {
```

Comment:
This doesn't seem logically correct to me. The engine internally deals with deleting markers. When a Back event is received, it should delete the last codepoint from the text store and if there is no content in the text store, it should pass the backspace on to the application so that the app can traverse text store boundaries (e.g. think bksp at start of a table cell -- moves to previous cell and deletes character, which is most likely a separate text store).

It looks like the kmx test case is making an unhelpful assumption here: I am updating test 020 to clarify start-of-context now.